File: python/src/GP_learning/transform_data.py
# -*- coding: utf-8 -*-
"""
Created on Wed Nov 27 14:10:37 2024

@author: rose
"""

import numpy as np

def transform_data(Data, method, transform_info=None):
    """
    Transforms the input data according to the specified method.
    If transform_info is provided, the transformation is based on that information.
    
    Parameters:
    Data (numpy.ndarray): The input data (shape should be [n_samples, n_features]).
    method (str): The transformation method ('none', 'center', 'zscore', 'rescale').
    transform_info (dict, optional): A dictionary with precomputed transformation information (default is None).
    
    Returns:
    Data_transformed (numpy.ndarray): The transformed data.
    transform_info (dict): Information used for the transformation (e.g., mean, std, min, max).
    """
    if method == 'none':  # No transformation
        Data_transformed = Data
        transform_info = {'method': 'none'}

    elif method == 'center':  # Center data (zero mean)
        mean_data = np.mean(Data, axis=0)
        Data_transformed = Data - mean_data
        transform_info = {'method': 'center', 'mean_data': mean_data}

    elif method == 'zscore':  # Z-score normalization (zero mean, unit variance)
        mean_data = np.mean(Data, axis=0)
        std_data = np.std(Data, axis=0)
        Data_transformed = (Data - mean_data) / std_data
        transform_info = {'method': 'zscore', 'mean_data': mean_data, 'std_data': std_data}

    elif method == 'rescale':  # Rescale data between -1 and 1
        min_data = np.min(Data, axis=0)
        max_data = np.max(Data, axis=0)
        a, b = -1, 1
        Data_transformed = a + ((Data - min_data) / (max_data - min_data)) * (b - a)
        transform_info = {'method': 'rescale', 'min_data': min_data, 'max_data': max_data, 'lb': a, 'ub': b}

    elif transform_info is not None:  # Transform using provided rules
        if transform_info['method'] == 'none':  # No transformation
            Data_transformed = Data

        elif transform_info['method'] == 'center':  # Center data (zero mean)
            Data_transformed = Data - transform_info['mean_data']

        elif transform_info['method'] == 'zscore':  # Z-score normalization (zero mean, unit variance)
            Data_transformed = (Data - transform_info['mean_data']) / transform_info['std_data']

        elif transform_info['method'] == 'rescale':  # Rescale data using provided min/max
            min_data = transform_info['min_data']
            max_data = transform_info['max_data']
            a, b = transform_info['lb'], transform_info['ub']
            Data_transformed = a + ((Data - min_data) / (max_data - min_data)) * (b - a)

    return Data_transformed, transform_info

File: matlab/LearnMPC_GPs/transform_data.m
%%
function [Data_transformed,transform_info] = transform_data(Data,method,varargin)
% see also normalize() function from matlab
% sadly this function does not have an inbuilt way to detransform the data
% again... so we do it here in an extra function

% the function can be used together with the function "detransform_data"

% if only "Data" and "method" are provided the function transforms the
% input data according to the method described in "method" (for example
% based on the mean of the provided data)

% if you provide a certain transformation rule transform_info =
% varargin{1}; 
%then data are transformed according to the provided rules (for example
%based on some provided mean value)

% its easier to understand if you directly look at the code

% the function outputs the transformed data and a transform_info struct,
% which stores all information to detransform the data or to transform new
% data point always in the same way. (if used as varargin{1} = transform_info)


if nargin == 2 % transform based on provided data
    if strcmp('none',method) % no transformation
        Data_transformed = Data;
        transform_info.method = 'none';

    end

    if strcmp('center',method) % data have zero mean
        sz = size(Data);
        mean_data = mean(Data);
        Data_transformed = Data - repmat(mean_data,sz(1),1);
        transform_info.method = 'center';
        transform_info.mean_data = mean_data;

    end

    if strcmp('zscore',method) % data have zero mean and std of 1
        sz = size(Data);
        mean_data = mean(Data);
        std_data = std(Data);
        mean_data_mat = repmat(mean_data,sz(1),1);
        std_data_mat = repmat(std_data,sz(1),1);
        Data_transformed = (Data - mean_data_mat)./std_data_mat;
        transform_info.method = 'zscore';
        transform_info.mean_data = mean_data;
        transform_info.std_data = std_data;

    end

    if strcmp('rescale',method)% for now scale data between -1 and 1 
        sz = size(Data);
        min_d = min(Data);
        max_d = max(Data);
        min_data = repmat(min_d,sz(1),1);
        max_data = repmat(max_d,sz(1),1);
        a = -1;
        b = 1;
        Data_transformed = a + ((Data-min_data)./(max_data-min_data)).*(b-a);
        transform_info.method = 'rescale';
        transform_info.min_data = min_d;
        transform_info.max_data = max_d;
        transform_info.lb = a;
        transform_info.ub = b;

    end

elseif nargin == 3 % transform according to provided rule
    transform_info = varargin{1};
    if strcmp('none',transform_info.method) % no transformation
        Data_transformed = Data;
        %transform_info.method = 'none';

    end

    if strcmp('center',transform_info.method) % data have zero mean
        sz = size(Data);
        Data_transformed = Data - repmat(transform_info.mean_data,sz(1),1);

    end

    if strcmp('zscore',transform_info.method) % data have zero mean and std of 1
        sz = size(Data);
        mean_data_mat = repmat(transform_info.mean_data,sz(1),1);
        std_data_mat = repmat(transform_info.std_data,sz(1),1);
        Data_transformed = (Data - mean_data_mat)./std_data_mat;

    end

    if strcmp('rescale',transform_info.method)% for now scale data between -1 and 1 
        sz = size(Data);
        min_data = repmat(transform_info.min_data,sz(1),1);
        max_data = repmat(transform_info.max_data,sz(1),1);
        a = transform_info.lb;%-1;
        b = transform_info.ub;% 1;
        Data_transformed = a + ((Data-min_data)./(max_data-min_data)).*(b-a);

    end



end %if nargin


end %fct